### ISSUE TITLE
Strange Behavior with Stream of Numbers

### ISSUE THREAD

**Issue by carloszimm** [mostjs/hold#31]
So, I came across a strange behavior while messing around with @most/hold. For the example, consider the following function which produces a infinite stream of numbers (starting with 0) according to a given time period:
```JavaScript
const interval = period => filter(x => x !== -1, scan((x, y) => x + y, -1, constant(1, skip(1, periodic(period)))));
```
Well, now, consider the following code:
```JavaScript
//let testHold = hold(interval(2000));
let testHold = multicast(interval(2000));

runEffects(tap(console.log, take(3, testHold)), newDefaultScheduler());

setTimeout(()=>{
    runEffects(tap(console.log, take(11, testHold)), newDefaultScheduler());
}, 4500);
```
With multicast, the code works as expected, printing: 0, 1, 2, 2, 3, 4, 5, ...
But, with hold, it acts strangely, printing: 0, 1, 1, 2, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6
I personally was expecting something more like 0, 1, 1, 2, 2, 3, 4, 5, ....
In other words, the second run should access the second emission (last from first run at time 4000 milliseconds), print one more number (2) together with the first run (the last number taken from the first run), and keep printing the reaming sequence.

Additionally, the program keeps hanging after the last two 6

**Comment by briancavalier** [comment]
Hey @cezl, sorry for the delay. I hope to have time to look at this over the weekend.

Cheers!

**Comment by carloszimm** [comment]
No problem, @briancavalier (:

**Comment by briancavalier** [comment]
@carloszimm Is there a real use case which led you to the test above?  If so, I think could be very helpful in understanding the issue.

The reason I ask is that the test is somewhat outside of defined behavior.  There are 3 sources of truth about time in the test, and there is no real defined relationship between them: two independent `@most/core` schedulers, and `setTimeout`.  Any observed relationship has to be treated as coincidence.  If you go digging into the implementation details of `@most/scheduler`, you'll find `setTimeout`, but it's just an implementation detail and not a formal part of the Scheduler abstraction and contract.

That said, this "extra" 2 is especially interesting:

> But, with hold, it acts strangely, printing: 0, 1, 1, 2, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6

I'll try it to see if I can reproduce the extra 2.  Let us know about the use cases.  Thanks!

**Comment by carloszimm** [comment]
Hey, @briancavalier 
> Is there a real use case which led you to the test above

Yes, there is. I was in fact using streams of numbers to test a functionality that I've been working on. I tried to mimic as close as possible in the above example, when a second stream is created (after a while - reason for having used a ```setTimeout```) that accesses the shared numeric stream. **I could use as well a stream generated directly by periodic, but I would not be able to see if the most recently seen event was being delivered correctly** :sweat_smile:

> That said, this "extra" 2 is especially interesting

It is, as well as the extra 3, 4, 5, and 6, since the first run stream should take only 3 numbers (if I in fact have understood how the hold operation actually works).

Still, there is the problem of the program not finishing.

> I'll try it to see if I can reproduce the extra 2

If possible, also alternate between multicast and hold to observe that it apparently works normally with multicast.

**Comment by briancavalier** [comment]
Thanks @carloszimm.

I reproduced the issue using your test case.  Thanks!  I'm pretty sure it's a bug related to the new flush algorithm that was introduced recently.  I'll keep you posted.

**Comment by briancavalier** [comment]
I just push a candidate fix to the `fix-pending-sink-management` branch in #32.  Feel free to give that branch a try, and let me know if it works for you.

**Comment by carloszimm** [comment]
> Thanks @carloszimm.

Yep (:

> Feel free to give that branch a try, and let me know if it works for you.

I've tested and it seems to be working as expected now :raised_hands:

Thank you, @briancavalier!

**Comment by briancavalier** [comment]
Great, thanks for trying it.  I should have time to add the test and release a patch version tonight (US East time).

**Comment by carloszimm** [comment]
Awesome! Looking forward.

**Comment by briancavalier** [comment]
Released 4.0.1 with the fix.  Thanks again for reporting the issue, for the test case, for and your patience!
